design/concise-tidy-design.Rmd

```diff
@@ -0,0 +1,215 @@
+
+### Principles
+The general principles I propose for tidying `crmPack` objects are as follows:
+
+* All slots that are not functions are converted to tibbles or a list of tibbles.
+* If the slot is scalar, the slot is converted to a 1x1 tibble.  This will ease downstream operations such as `row-bind`ing.
+* If the object being tidied contains multiple slots of (potentially) different lengths, the results is a list of tibbles.  The list may be nested to multiple levels.  (See, for example, `LogisticLogNormal`.)
+* The column names of the tidied tibble correspond to the slot names of the parent object.
+* If an object contains a vector or matrix of values that correspond to an "obvious" set of parameters, columns take on the names of those parameters and the tibble is augmented by a column named `Parameter`.  This most often happens in the mean vector and covariance matrix of `ModelParamsNormal`.
+* When the value of a slot has not been set, a zero-row tibble is returned.
+* When the value of a slot has scalar attributes, these attributes are added as a column of the tibble, whose name is the name of the attribute and whose value is the value of the attribute for every row of the tibble.  Vector attributes ill have to be handled on a case-by-case basis.
+* `tbl_<className>` is prepended to the class of the tidy tibble.
+
+### Exceptions
+*  Where a vector slot (or series of vector slots) define a range, then the naming convention described above is not followed.  Instead, columns named `min` and `max` define the extent of the range and a column named `Range` labels the range.
+
+### Please also read
+Please also read [Issue 407](https://github.com/Roche/crmPack/issues/407) regarding inconsistent slot names.
+
+## Implementation
+`tidy()` is a generic defined in `broom`.  An end-user of `crmPack` may not have installed `broom`.  Therefore, we need to define an appropriate geneic when necessary.
+
+### Definition of the generic
+```{r}
+if (!requireNamespace("broom", quietly = TRUE)) {
+  #' Conversion of crmPack objects to data.frames/tibbles
+  #'
+  #' @description `r lifecycle::badge("experimental")`
+  #'
+  #' A method that converts a crmPack S4 object to a tibble or list of tibbles,
+  #' in the style of `broom::tidy()`.
+  #'
+  #' @param object \cr the object to convert
+  #' @param ... additional arguments passed to class-specific implementations.
+  #' @return (`tbl_df` or named `list` [of `tbl_df`])
+  #' @export
+  #'
+  setGeneric(
+    name = "tidy",
+    def = function(object, ...) {
+      standardGeneric("tidy")
+    },
+    valueClass = c("tbl_df", "tbl", "data.frame", "list")
+  )
+} else {
+  library(broom)
+}
+```
+
+### Simple examples
+
+`CohortSizeConst` is a trivial example.
+
+```{r}
+# tidy-CohortSize ----
+
+#' @rdname tidy
+#' @aliases tidy-CohortSize
+#' @example examples/CohortSize-method-tidy.R
+#' @export
+setMethod(
```

Comment:
this could be a default method, defined with a signature for the top level classes, because the content is not specific

Reply:
Yes, and then that would satisfy your request for a "do it even if it's ugly" default method.

Reply:
yep, good